Add upper bounds to fix 0.2.3 build

jdblischak opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

---

It's difficult to troubleshoot the failing update PRs when the currently merged 0.2.3 is also failing. This is due to multiple breaking changes in its dependencies. This PR adds the upper bounds. We'll also need repodata patches to patch the existing spatialdata conda binaries.

Here is the motivation behind all the upper bounds:

**dask-core**

Needs to be `<2024.11.0` because 2024.11.0 deprecated the legacy DataFrame backend
(https://github.com/dask/dask/pull/11606#issuecomment-2606834431). Without this upper bound, `import spatialdata` fails with `NotImplementedError: The legacy implementation is no longer supported`

**ome-zarr**

This is a conda-forge specific problem. ome-zarr-py 0.10.3 introduced setuptools-scm, but the feedstock was not updated to accommodate this (https://github.com/conda-forge/ome-zarr-feedstock/issues/24). Thus `pip check` fails with `spatialdata 0.2.3 has requirement ome-zarr>=0.8.4, but you have ome-zarr 0.0.0.`

**xarray**

Found this upper bound empirically based on when spatialdata 0.2.3 was merged into this feedstock in Sep 2024 (https://github.com/conda-forge/spatialdata-feedstock/commit/7186ce99100255baea2c85359914109b0de36ff2). I confirmed locally that I could `import spatialdata` with xarray 2024.09 but that it fails with 2024.10 with the error `ImportError: cannot import name 'HybridMappingProxy' from 'xarray.core.utils'`

jdblischak committed:
MNT: Re-rendered with conda-build 25.1.2, conda-smithy 3.47.0, and conda-forge-pinning 2025.03.10.15.44.50

jdblischak committed:
Add upper bound on xarray b/c 2024.10 is incompatible

jdblischak committed:
Add upper bound on dask b/c 2024.11.0 deprecated the legacy DataFrame backend

jdblischak committed:
Add upper bound on ome-zarr b/c 0.10.3 reports version 0.0.0

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

I do have some suggestions for making it better though...


For **recipe/meta.yaml**:

 * ℹ️ `noarch: python` recipes should usually follow the syntax in our [documentation](https://conda-forge.org/docs/maintainer/knowledge_base/#noarch-python) for specifying the Python version.
   - For the `host` section of recipe, you should usually use `python {{ python_min }}` for the `python` entry.
   - For the `run` section of recipe, you should usually use `python >={{ python_min }}` for the `python` entry.
   - For the `test.requires` section of recipe, you should usually use `python {{ python_min }}` for the `python` entry.
   - If the package requires a newer Python version than the currently supported minimum version on `conda-forge`, you can override the `python_min` variable by adding a Jinja2 `set` statement at the top of your recipe (or using an equivalent `context` variable for v1 recipes).
 * ℹ️ PyPI default URL is now pypi.org, and not pypi.io. You may want to update the default source url.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/13772012276](https://github.com/conda-forge/conda-forge-webservices/actions/runs/13772012276). Examine the logs at this URL for more detail.</sub>

LucaMarconato commented on recipe/meta.yaml:
```diff
@@ -23,15 +23,15 @@ requirements:
     - anndata >=0.9.1
     - click
-    - dask-core >=2024.4.1
+    - dask-core >=2024.4.1,<2024.11.0
```
Looks good to me. The latest version of `spatialdata` works up to `2024.12.1`, so this upper bound could be used. But since `spatialdata==0.2.4` was released in November, it's also an option to use the upper bound you specified for `0.2.3` and then relax it with `0.2.4`.

jdblischak replied:
> The latest version of `spatialdata` works up to `2024.12.1`, so this upper bound could be used. 

Thanks for the extra context

> But since `spatialdata==0.2.4` was released in November, it's also an option to use the upper bound you specified for `0.2.3` and then relax it with `0.2.4`.

Right, my plan is to go step-by-step. First get 0.2.3 functional again. Then release 0.2.5. And then finally 0.3.0 (which as you pointed out in that PR is blocked by xarray at the moment anyways). I'll update the lower and upper bounds as I go

LucaMarconato replied:
Thanks a lot!

LucaMarconato commented:
Hi, looks great! An extra comment: the `pyproject.toml` in `spatialdata` at the time of the release also includes explicitly `scipy` and `scikit-image`. I would therefore add these to `meta.yaml` as well.

jdblischak commented:
I'm going to fix the lint first. Then I'll merge

LucaMarconato commented:
Ah no, they are already there, I did a mistake when checking the diff. So good to me.

jdblischak committed:
Apply CFEP-25 to set python_min to 3.9

Confirmed that 0.2.3 has `requires-python = ">=3.9"`

https://github.com/scverse/spatialdata/blob/v0.2.3/pyproject.toml#L18

jdblischak committed:
MNT: Re-rendered with conda-build 25.1.2, conda-smithy 3.47.0, and conda-forge-pinning 2025.03.10.16.49.26

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

jdblischak commented:
> **dask-core**
>
> Needs to be `<2024.11.0` because 2024.11.0 deprecated the legacy DataFrame backend
(https://github.com/dask/dask/pull/11606#issuecomment-2606834431). Without this upper bound, `import spatialdata` fails with `NotImplementedError: The legacy implementation is no longer supported`

Quick follow-up to document that @LucaMarconato was correct to comment on the upper bound above in https://github.com/conda-forge/spatialdata-feedstock/pull/17#discussion_r1987834615 and in the 0.2.5 PR in https://github.com/conda-forge/spatialdata-feedstock/pull/10#issuecomment-2714089367. I was too focused on when the legacy DataFrame implementation was deprecated in [2024.11.0](https://docs.dask.org/en/stable/changelog.html#v2024-11-2), which only generates a warning, and and when it was actually removed in [2025.1.0](https://docs.dask.org/en/stable/changelog.html#v2025-1-0), which generates an error.

I relaxed the upper bound to `<2025` for the 0.2.5 conda binary. I will also use that upper bound for the repodata patch I plan to submit later this week. However, I am going to leave the 0.2.3 build number 1 conda binary generated by this PR as is since anyone using an older spatialdata should be fine using an older dask as well.

Documenting that spatialdata 0.2.3 installed with dask 2024.12.1 only generates a warning:

```sh
python -m venv ./spatial
source ./spatial/bin/activate
python --version
## Python 3.10.16
pip install --prefer-binary spatialdata==0.2.3 'dask<2025' multiscale-spatial-image==1 xarray==2024.09

pip check
## No broken requirements found.
pip list | grep dask
## dask                     2024.12.1
## dask-expr                1.1.21
## dask-image               2024.5.3

python -c "import spatialdata"
## /home/wsl/repos/spatialdata-feedstock/spatial/lib/python3.10/site-packages/dask/dataframe/__init__.py:31: FutureWarning: The legacy Dask DataFrame implementation is deprecated and will be removed in a future version. Set the configuration option `dataframe.query-planning` to `True` or None to enable the new Dask Dataframe implementation and silence this warning.
##  warnings.warn(
```

jdblischak commented:
> We'll also need repodata patches to patch the existing spatialdata conda binaries.

I finally got around to submitting the repodata patch in https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/1041

jdblischak commented:
> I finally got around to submitting the repodata patch in https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/1041

It was merged (https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/commit/467ffadd6fee4b8d3c30c832a7ef58cabe451a4b). I confirmed the patches work as expected:

```sh
mamba create --dry-run -n test -c conda-forge \
  spatialdata==0.2.3 dask-core=2025
## error    libmamba Could not solve for environment specs
##     The following packages are incompatible
##     ├─ dask-core =2025 * is requested and can be installed;
##     └─ spatialdata ==0.2.3 * is not installable because there are no viable options
##        ├─ spatialdata 0.2.3 would require
##        │  └─ dask-core >=2024.4.1,<2025.0.0a0 *, which conflicts with any installable versions previously reported;
##        └─ spatialdata 0.2.3 would require
##           └─ dask-core >=2024.4.1,<2024.11.0 *, which conflicts with any installable versions previously reported.
## critical libmamba Could not solve for environment specs
mamba create --dry-run -n test -c conda-forge \
  spatialdata==0.2.3 xarray="2024.10"
## error    libmamba Could not solve for environment specs
##     The following packages are incompatible
##     ├─ spatialdata ==0.2.3 * is installable and it requires
##     │  └─ xarray [>=2023.12.0,<2024.10 *|>=2023.12.0,<2024.10.0a0 *], which can be installed;
##     └─ xarray =2024.10 * is not installable because it conflicts with any installable versions previously reported.
## critical libmamba Could not solve for environment specs
```
